File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/rooms/android/RoomsGameActivity.java
package com.zwstudio.logicpuzzlesandroid.puzzles.rooms.android;

import android.view.View;

import com.zwstudio.logicpuzzlesandroid.R;
import com.zwstudio.logicpuzzlesandroid.common.android.GameActivity;
import com.zwstudio.logicpuzzlesandroid.common.data.MoveProgress;
import com.zwstudio.logicpuzzlesandroid.puzzles.rooms.data.RoomsDocument;
import com.zwstudio.logicpuzzlesandroid.puzzles.rooms.domain.RoomsGame;
import com.zwstudio.logicpuzzlesandroid.puzzles.rooms.domain.RoomsGameMove;
import com.zwstudio.logicpuzzlesandroid.puzzles.rooms.domain.RoomsGameState;

import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.EActivity;

import java.util.List;

@EActivity(R.layout.activity_game_game)
public class RoomsGameActivity extends GameActivity<RoomsGame, RoomsDocument, RoomsGameMove, RoomsGameState> {
    public RoomsDocument doc() {return app.roomsDocument;}

    protected RoomsGameView gameView;
    protected View getGameView() {return gameView;}

    @AfterViews
    protected void init() {
        gameView = new RoomsGameView(this);
        super.init();
    }

    protected void startGame() {
        String selectedLevelID = doc().selectedLevelID;
        List<String> layout = doc().levels.get(selectedLevelID);
        tvLevel.setText(selectedLevelID);
        updateSolutionUI();

        levelInitilizing = true;
        game = new RoomsGame(layout, this);
        try {
            // restore game state
            for (MoveProgress rec : doc().moveProgress()) {
                RoomsGameMove move = doc().loadMove(rec);
                game.setObject(move);
            }
            int moveIndex = doc().levelProgress().moveIndex;
            if (!(moveIndex >= 0 && moveIndex < game.moveCount())) return;
            while (moveIndex != game.moveIndex())
                game.undo();
        } finally {
            levelInitilizing = false;
        }
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/rooms/RoomsGameActivity.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.rooms

import com.zwstudio.logicpuzzlesandroid.R
import com.zwstudio.logicpuzzlesandroid.common.android.GameGameActivity
import com.zwstudio.logicpuzzlesandroid.common.data.GameLevel
import org.androidannotations.annotations.AfterViews
import org.androidannotations.annotations.Bean
import org.androidannotations.annotations.Click
import org.androidannotations.annotations.EActivity

@EActivity(R.layout.activity_game_game)
class RoomsGameActivity : GameGameActivity<RoomsGame, RoomsDocument, RoomsGameMove, RoomsGameState>() {
    @Bean
    protected lateinit var document: RoomsDocument
    override val doc get() = document

    @AfterViews
    protected override fun init() {
        gameView = RoomsGameView(this)
        super.init()
    }

    override fun newGame(level: GameLevel) =
        RoomsGame(level.layout, this, doc)

    @Click
    protected fun btnHelp() {
        RoomsHelpActivity_.intent(this).start()
    }
}